utils/utils.go

```diff
@@ -51,32 +53,67 @@ func eofKeySequenceText() string {
 }
 
 // AddSecret Recalls state if present, and appends a secret to the state file
-func AddSecret(toAdd Secret) {
-	ism := NewIOStateManager(defaultFile)
-	secrets, _ := ism.ReadState()
+func AddSecret(toAdd Secret, keyURI string, keyCheck bool) {
+	envelope := V2{Filepath: defaultFile}
+	envelope.KeyIdentifier = keyURI
+	nvl := NewVersionedLoader(defaultFile)
+
+	contents, err := nvl.ReadState()
+	// First run case with FileNotFound exception. Mask this and save
+	if os.IsNotExist(err) {
+		envelope.Secrets.Add(toAdd)
+		envelope.Save()
+		return
+	}
 
-	secrets.Add(toAdd)
-	ism.WriteState(secrets)
+	// Keycheck overrides if we bother with the key evaluation. This is problematic
+	// when doing re-key operations on a post-v2 migrated envelope.
+	if keyCheck {
+		// If we load contents and the user has a different keyURI, intercept
+		// and warn before they save a multi-key sealed envelope of secrets
+		if !contents.SameKey(keyURI) {
+			log.Fatalf("Key mismatch detected, Refusing to save with different key identifiers. Envelope Provided: %s  User Provided: %s", contents.KeyIdentifier, keyURI)
+		}
+	}
+
+	// Handle any other parsing errors in a Fatal fashion
+	if err != nil {
+		log.WithFields(log.Fields{"method": "AddSecret"}).Fatal("Failed parsing all known envelope formats.")
+	}
+
+	jsonData, _ := json.MarshalIndent(envelope, "", " ")
+	log.Debug(string(jsonData))
+	envelope.Secrets = contents.Secrets
+
+	envelope.Secrets.Add(toAdd)
+	envelope.Save()
 }
 
 // ReadSecrets is a Wrapper to return an array of Secrets for processing
-func ReadSecrets() Secrets {
-	ism := NewIOStateManager(defaultFile)
-	secrets, err := ism.ReadState()
+// Along with any KeyIdentifier found in the envelope for decryption
+func ReadSecrets() (Secrets, string, error) {
+	nvl := NewVersionedLoader(defaultFile)
+
+	contents, err := nvl.ReadState()
 	if err != nil {
-		log.Fatalf("Error reading state - %s", err)
+		log.WithFields(log.Fields{"method": "ReadSecrets"}).Warn("Failed parsing all known envelope formats.")
 	}
-	return secrets
+	return contents.Secrets, contents.KeyIdentifier, nil
 }
 
 // DeleteSecret is a Wrapper to remove a secret from state
 // toRemove - string - named key of the secret to eject from the state storage
 func DeleteSecret(toRemove string) {
-	ism := NewIOStateManager(defaultFile)
-	secrets, err := ism.ReadState()
+	nvl := NewVersionedLoader(defaultFile)
+
+	contents, err := nvl.ReadState()
 	if err != nil {
-		log.Fatalf("Error reading state - %s", err)
+		log.WithFields(log.Fields{"method": "DeleteSecret"}).Warn("Failed parsing all known envelope formats.")
```

Comment:
should this just be a warning? I feel like ignoring the error here could potentially make things weird, since ReadState will return an empty V2 on error, so then wee are saving that empty state, overwriting what may have been there before